File: sdk/reproit-ios/Sources/ReproIt/CaptureBatch.swift
// capture-batch-v1 emission for iOS production failure occurrences.
//
// Port of the batch builder in sdk/reproit-backend-node/capture.js. A failure
// that carries recorded dependency exchanges ships as a source-neutral causal
// capture batch (`POST <endpoint>/v1/capture-batches`) so the occurrence can
// be re-executed hermetically instead of only re-evaluated. The legacy event
// batch (`/v1/events`) is untouched and still carries every other event.
//
// Event order mirrors the backend SDKs exactly:
//   operation-start, trigger, determinism-envelope checkpoint,
//   dependency (one per exchange), operation-end, observation.

import Foundation

/// Protocol token charset (`validate_token` in reproit-protocol).
func reproitIsProtocolToken(_ value: String?) -> Bool {
  guard let value, !value.isEmpty, value.count <= 128 else { return false }
  return value.allSatisfy { character in
    character.isASCII
      && (character.isLetter || character.isNumber || "._:-".contains(character))
  }
}

/// Builds one capture-batch-v1 for a failure occurrence.
enum ReproItCaptureBatch {
  /// The emitter identity for this SDK, used in every event id.
  static let emitterId = "mobile-ios"

  /// Code identity, in priority order: explicit config, then the common CI and
  /// platform environment. Never shells out to git.
  static func resolveCommit(_ configured: String?) -> String? {
    let environment = ProcessInfo.processInfo.environment
    for candidate in [configured, environment["REPROIT_COMMIT"], environment["GITHUB_SHA"]] {
      if reproitIsProtocolToken(candidate) { return candidate }
    }
    return nil
  }

  /// The determinism envelope: where and when the capture happened, and a seed
  /// that makes REPLAY runs deterministic. Honesty note, carried from the
  /// backend SDKs: the seed does not reproduce the randomness the app drew in
  /// production; it pins the replay's.
  static func envelopeAttributes(observedAtMs: Int64, replaySeed: String? = nil) -> [String: Any] {
    var attributes: [String: Any] = [
      "observedAtMs": observedAtMs,
      "tz": TimeZone.current.identifier,
      "runtime": "swift",
      "os": reproitPlatformName,
      "arch": reproitDeviceArch,
      "replaySeed": replaySeed ?? reproitRandomSeedHex(),
    ]
    let environment = ProcessInfo.processInfo.environment
    if reproitIsProtocolToken(environment["REPROIT_IMAGE_DIGEST"]) {
      attributes["imageDigest"] = environment["REPROIT_IMAGE_DIGEST"]!
    }
    if let build = Bundle.main.infoDictionary?["CFBundleVersion"] as? String,
      reproitIsProtocolToken(build)
    {
      attributes["buildDigest"] = build
    }
    return attributes
  }

  /// Build the batch. Returns nil when the occurrence carries no exchanges:
  /// without them a capsule cannot be re-executed, so claiming the network
  /// capability would be an over-claim and the legacy path already covers the
  /// evidence-only case.
  ///
  /// - Parameters:
  ///   - appId: the cloud project id.
  ///   - sessionId: the SDK session identity (also the batch's trace id).
  ///   - operation: the failing operation name (the screen signature).
  ///   - triggerSubject: what the user did to reach it.
  ///   - triggerValue: the replayable trigger input, already redacted.
  ///   - exchanges: production-recorded dependency exchanges, oldest first.
  ///   - failureSummary: human-readable failure text.
  ///   - failureSignature: the grouping signature.
  static func build(
    appId: String,
    sessionId: String,
    batchId: String,
    operation: String,
    triggerSubject: String,
    triggerValue: Any?,
    exchanges: [ReproItExchange],
    failureSummary: String,
    failureSignature: String,
    buildVersion: String?,
    buildCommit: String?,
    observedAtMs: Int64 = reproitNowMs(),
    replaySeed: String? = nil
  ) -> [String: Any]? {
    guard !exchanges.isEmpty else { return nil }

    var events: [[String: Any]] = []
    var sequence = 0
    var previousId: String?

    func append(_ event: [String: Any]) {
      sequence += 1
      let id = "evt_\(emitterId)_\(sequence)"
      var captured: [String: Any] = [
        "id": id,
        "sequence": sequence,
        // Monotonic offsets are the event ordinal here: the mobile capture
        // path has no per-event high-resolution clock, and inventing one
        // would be a fabricated envelope.
        "monotonicNs": sequence,
        "causalParentIds": previousId.map { [$0] } ?? [],
        "traceId": sessionId,
        "event": event,
      ]
      captured["actor"] = nil
      captured.removeValue(forKey: "actor")
      events.append(captured)
      previousId = id
    }

    append(["kind": "operation-start", "name": operation])
    let value: [String: Any] =
      triggerValue == nil
      ? ["representation": "structural", "shape": ["type": "unknown"]]
      : [
        "representation": "replayable",
        "value": reproitRedactExchangeValue(triggerValue!),
        "redaction": "redacted-at-source",
      ]
    append([
      "kind": "trigger", "trigger": "ui-action", "subject": triggerSubject, "value": value,
    ])
    append([
      "kind": "checkpoint",
      "name": "determinism-envelope",
      "attributes": envelopeAttributes(observedAtMs: observedAtMs, replaySeed: replaySeed),
    ])
    for exchange in exchanges {
      append([
        "kind": "dependency",
        "system": "service",
        "operation": "call",
        "subject": exchange.url,
        "value": [
          "representation": "replayable",
          "value": exchange.jsonObject(),
          "redaction": "redacted-at-source",
        ],
      ])
    }
    append(["kind": "operation-end", "name": operation, "outcome": "failed"])
    append([
      "kind": "observation",
      "failure": [
        "observation": "exception",
        "authority": "runtime-diagnosis",
        "summary": failureSummary,
        "signature": failureSignature,
        "observationPoint": operation,
        "artifactIds": [] as [String],
      ],
    ])

    var batch: [String: Any] = [
      "version": 1,
      "batchId": batchId,
      "projectId": appId,
      "sessionId": sessionId,
      "emitter": [
        "id": emitterId,
        "kind": "runtime-sdk",
        "component": "mobile",
        "runtime": "swift",
      ],
      "observedAt": reproitIso8601(observedAtMs),
      "policy": ["consent": "application-telemetry", "retentionClass": "standard"],
      "capabilities": [
        [
          "capability": "network",
          "completeness": "complete",
          "detail": "outbound dependency exchanges recorded with responses",
        ]
      ],
      "events": events,
      "artifacts": [] as [Any],
    ]
    let commit = resolveCommit(buildCommit)
    var deployment: [String: Any] = [:]
    if reproitIsProtocolToken(buildVersion) { deployment["version"] = buildVersion! }
    if let commit { deployment["commit"] = commit }
    if !deployment.isEmpty { batch["deployment"] = deployment }
    return batch
  }
}

/// Device architecture reported in the envelope.
let reproitDeviceArch: String = {
  #if arch(arm64)
    return "arm64"
  #elseif arch(x86_64)
    return "x86_64"
  #else
    return "unknown"
  #endif
}()

/// 8 random bytes as lowercase hex, matching the backend SDKs' replay seed.
func reproitRandomSeedHex() -> String {
  (0..<8).map { _ in String(format: "%02x", UInt8.random(in: 0...255)) }.joined()
}

/// RFC 3339 timestamp for `observedAt`.
func reproitIso8601(_ millis: Int64) -> String {
  let formatter = DateFormatter()
  formatter.locale = Locale(identifier: "en_US_POSIX")
  formatter.timeZone = TimeZone(secondsFromGMT: 0)
  formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
  return formatter.string(from: Date(timeIntervalSince1970: Double(millis) / 1000.0))
}

File: sdk/reproit-android/src/main/kotlin/com/reproit/android/CaptureBatch.kt
package com.reproit.android

import java.util.Locale
import java.util.TimeZone

/**
 * Universal causal capture batch (`capture-batch-v1`), the Android port of
 * `sdk/reproit-recorder-node/index.js` plus the batch assembly in
 * `sdk/reproit-backend-node/capture.js`.
 *
 * A failure occurrence that carries recorded dependency exchanges ships on this
 * contract (`POST <endpoint>/v1/capture-batches`) instead of the legacy event
 * batch, because only this shape survives the projection back into a replayable
 * capture. Dependency exchanges are conditional evidence.
 *
 * Pure Kotlin with no `android.*` import, so the emitted batch is host-testable
 * against the protocol validator.
 */

/** The ingest protocol token charset (`validate_token` in reproit-protocol). */
private val TOKEN = Regex("^[A-Za-z0-9._:-]{1,128}$")

internal fun validToken(value: String?): Boolean = value != null && TOKEN.matches(value)

/** Recorder bounds. The protocol allows more; these keep a mobile batch small. */
private const val MAX_BATCH_EVENTS = 1024

/**
 * Incremental recorder for one occurrence. Mirrors the Node recorder's event
 * identity (`evt_<emitter>_<sequence>`), monotonic stamping, and causal parent
 * chaining exactly, so a batch from Android is indistinguishable in shape from
 * one emitted by the backend SDKs.
 */
internal class CaptureRecorder(
  private val batchId: String,
  private val projectId: String,
  private val sessionId: String,
  private val emitterId: String,
  private val emitterComponent: String,
  private val emitterRuntime: String,
  private val observedAt: String,
  private val deployment: Map<String, Any?>?,
  private val capabilities: List<Map<String, Any?>>,
) {
  private val events = ArrayList<Map<String, Any?>>()
  private var sequence = 1
  private var finished = false

  /** Record one event and return its id, or null once the batch is bounded out. */
  fun record(
    event: Map<String, Any?>,
    parent: String? = null,
    traceId: String? = null,
    monotonicNs: Long? = null,
  ): String? {
    if (finished || events.size >= MAX_BATCH_EVENTS) return null
    val ordinal = sequence++
    val id = "evt_${emitterId}_$ordinal"
    val captured = LinkedHashMap<String, Any?>()
    captured["id"] = id
    captured["sequence"] = ordinal
    captured["monotonicNs"] = monotonicNs ?: ordinal.toLong()
    captured["causalParentIds"] = if (parent == null) emptyList<String>() else listOf(parent)
    if (traceId != null) captured["traceId"] = traceId
    captured["event"] = event
    events.add(captured)
    return id
  }

  fun operationStart(name: String, parent: String?, traceId: String?, monotonicNs: Long?): String? =
    record(linkedMapOf("kind" to "operation-start", "name" to name), parent, traceId, monotonicNs)

  fun trigger(
    trigger: String,
    subject: String,
    value: Map<String, Any?>?,
    parent: String?,
    traceId: String?,
    monotonicNs: Long?,
  ): String? {
    val event = linkedMapOf<String, Any?>("kind" to "trigger", "trigger" to trigger)
    event["subject"] = subject
    if (value != null) event["value"] = value
    return record(event, parent, traceId, monotonicNs)
  }

  fun checkpoint(
    name: String,
    attributes: Map<String, Any?>,
    parent: String?,
    traceId: String?,
    monotonicNs: Long?,
  ): String? =
    record(
      linkedMapOf("kind" to "checkpoint", "name" to name, "attributes" to attributes),
      parent,
      traceId,
      monotonicNs,
    )

  fun dependency(
    system: String,
    operation: String,
    subject: String,
    value: Map<String, Any?>?,
    parent: String?,
    traceId: String?,
    monotonicNs: Long?,
  ): String? {
    val event = linkedMapOf<String, Any?>("kind" to "dependency", "system" to system)
    event["operation"] = operation
    event["subject"] = subject
    if (value != null) event["value"] = value
    return record(event, parent, traceId, monotonicNs)
  }

  fun operationEnd(
    name: String,
    outcome: String,
    parent: String?,
    traceId: String?,
    monotonicNs: Long?,
  ): String? =
    record(
      linkedMapOf("kind" to "operation-end", "name" to name, "outcome" to outcome),
      parent,
      traceId,
      monotonicNs,
    )

  fun failure(
    failure: Map<String, Any?>,
    parent: String?,
    traceId: String?,
    monotonicNs: Long?,
  ): String? = record(linkedMapOf("kind" to "observation", "failure" to failure), parent, traceId, monotonicNs)

  fun finish(): Map<String, Any?> {
    finished = true
    val batch = LinkedHashMap<String, Any?>()
    batch["version"] = 1
    batch["batchId"] = batchId
    batch["projectId"] = projectId
    batch["sessionId"] = sessionId
    batch["emitter"] =
      linkedMapOf(
        "id" to emitterId,
        "kind" to "runtime-sdk",
        "component" to emitterComponent,
        "runtime" to emitterRuntime,
      )
    if (deployment != null) batch["deployment"] = deployment
    batch["observedAt"] = observedAt
    batch["policy"] =
      linkedMapOf("consent" to "application-telemetry", "retentionClass" to "standard")
    batch["capabilities"] = capabilities
    batch["events"] = events
    batch["artifacts"] = emptyList<Any?>()
    return batch
  }
}

/** A replayable captured value in the protocol's representation vocabulary. */
internal fun replayableValue(value: Any?): Map<String, Any?> =
  linkedMapOf(
    "representation" to "replayable",
    "value" to value,
    "redaction" to "redacted-at-source",
  )

/** A structural captured value: shape retained, contents deliberately absent. */
internal fun structuralValue(shape: Any?): Map<String, Any?> =
  linkedMapOf("representation" to "structural", "shape" to shape)

/** One recorded outbound exchange with its ordering and subject metadata. */
internal data class CapturedExchange(
  val subject: String,
  val exchange: Map<String, Any?>,
  val atMs: Long,
  val monotonicNs: Long,
)

/**
 * The determinism envelope: where and when the capture happened, plus a seed
 * that makes REPLAY runs deterministic. Honesty note carried from the backend
 * SDKs: the seed does not reproduce the randomness the app drew in production,
 * it pins the replay's.
 */
internal fun determinismEnvelope(
  observedAtMs: Long,
  osRelease: String,
  arch: String,
  replaySeed: String,
  imageDigest: String?,
): MutableMap<String, Any?> {
  val envelope = LinkedHashMap<String, Any?>()
  envelope["observedAtMs"] = observedAtMs
  envelope["tz"] = TimeZone.getDefault().id
  envelope["runtime"] = "android"
  envelope["os"] = osRelease
  envelope["arch"] = arch
  envelope["replaySeed"] = replaySeed
  if (validToken(imageDigest)) envelope["imageDigest"] = imageDigest
  return envelope
}

/**
 * Assemble one failure occurrence into a capture batch: the UI action that
 * triggered it, the determinism envelope, every recorded dependency exchange in
 * order, and the failure observation carrying the oracle signature.
 *
 * Returns null when the identifiers the ingest protocol requires are unusable,
 * so a misconfigured app stops locally instead of emitting a batch the server
 * will reject.
 */
internal fun buildFailureCaptureBatch(
  appId: String,
  sessionId: String,
  operation: String,
  triggerAction: String,
  triggerValue: Any? = null,
  signature: String,
  summary: String,
  observationPoint: String,
  exchanges: List<CapturedExchange>,
  envelope: Map<String, Any?>,
  buildVersion: String?,
  buildCommit: String?,
  observedAtIso: String,
  batchSequence: Long,
  observedAtMs: Long,
): Map<String, Any?>? {
  if (!validToken(appId)) return null
  val batchId = "cb-android-$observedAtMs-$batchSequence"
  if (!validToken(batchId)) return null
  val deployment = LinkedHashMap<String, Any?>()
  if (validToken(buildVersion)) deployment["version"] = buildVersion
  if (validToken(buildCommit)) deployment["commit"] = buildCommit
  val capabilities =
    mutableListOf<Map<String, Any?>>(
      linkedMapOf("capability" to "user-interface", "completeness" to "complete")
    )
  // Declared only when exchanges were actually recorded, so the capsule
  // completeness model never over-claims on a capture without them.
  if (exchanges.isNotEmpty()) {
    capabilities.add(
      linkedMapOf(
        "capability" to "network",
        "completeness" to "complete",
        "detail" to "outbound dependency exchanges recorded with responses",
      )
    )
  }
  val recorder =
    CaptureRecorder(
      batchId = batchId,
      projectId = appId,
      sessionId = sessionId,
      emitterId = "mobile-android",
      emitterComponent = "mobile",
      emitterRuntime = "android",
      observedAt = observedAtIso,
      deployment = if (deployment.isEmpty()) null else deployment,
      capabilities = capabilities,
    )
  var parent = recorder.operationStart(operation, null, sessionId, observedAtMs)
  parent =
    recorder.trigger(
      "ui-action",
      operation,
      replayableValue(triggerValue ?: linkedMapOf("action" to triggerAction)),
      parent,
      sessionId,
      observedAtMs,
    )
  parent = recorder.checkpoint("determinism-envelope", envelope, parent, sessionId, observedAtMs)
  for (captured in exchanges) {
    parent =
      recorder.dependency(
        "service",
        "call",
        captured.subject,
        replayableValue(captured.exchange),
        parent,
        sessionId,
        captured.monotonicNs,
      )
  }
  parent = recorder.operationEnd(operation, "failed", parent, sessionId, observedAtMs)
  recorder.failure(
    linkedMapOf(
      "observation" to "exception",
      "authority" to "runtime-diagnosis",
      "summary" to summary,
      "signature" to signature,
      "observationPoint" to observationPoint,
      "artifactIds" to emptyList<String>(),
    ),
    parent,
    sessionId,
    observedAtMs,
  )
  return recorder.finish()
}

/** A bounded hex seed for the determinism envelope. */
internal fun replaySeedHex(random: () -> Long): String =
  java.lang.Long.toHexString(random()).padStart(16, '0').lowercase(Locale.ROOT).takeLast(16)
